lib/src/git.rs

```diff
@@ -2145,6 +2155,157 @@ pub fn expand_fetch_refspecs(
+                Instruction::Fetch(fetch) => match fetch {
+                    gix::refspec::instruction::Fetch::Only { src: _ } => {
+                        ignored_refspecs.push(IgnoredRefspec {
+                            refspec: refspec.to_bstring(),
+                            reason: "fetch-only refspecs are not supported",
+                        });
+                        return None;
+                    }
+
+                    // TODO: these need to be included, but we currently have no way to express
+                    // negative refspecs
+                    gix::refspec::instruction::Fetch::Exclude { src: _ } => return None,
+
+                    gix::refspec::instruction::Fetch::AndUpdate {
+                        src,
+                        dst,
+                        allow_non_fast_forward: _, // Already captured above
+                    } => (ensure_utf8(src)?, ensure_utf8(dst)?),
+                },
+            };
+
+            if !forced {
+                ignored_refspecs.push(IgnoredRefspec {
+                    refspec: refspec.to_bstring(),
+                    reason: "non-forced refspecs are not supported",
+                });
+                return None;
+            }
+
+            let Some(src_branch) = src.strip_prefix("refs/heads/") else {
+                ignored_refspecs.push(IgnoredRefspec {
+                    refspec: refspec.to_bstring(),
+                    reason: "only refs/heads/ is supported for refspec sources",
+                });
+                return None;
+            };
+
+            let dst = {
+                let Some(dst_without_prefix) = dst.strip_prefix("refs/remotes/") else {
+                    ignored_refspecs.push(IgnoredRefspec {
+                        refspec: refspec.to_bstring(),
+                        reason: "only refs/remotes/ is supported for fetch destinations",
+                    });
+                    return None;
+                };
+
+                let Some(dst_branch) = dst_without_prefix
+                    .strip_prefix(remote_name)
+                    .and_then(|d| d.strip_prefix("/"))
+                else {
+                    ignored_refspecs.push(IgnoredRefspec {
+                        refspec: refspec.to_bstring(),
+                        reason: "remote renaming not supported",
+                    });
+                    return None;
+                };
+
+                if src_branch == dst_branch {
+                    dst.to_owned()
+                } else {
+                    ignored_refspecs.push(IgnoredRefspec {
+                        refspec: refspec.to_bstring(),
+                        reason: "renaming is not supported",
+                    });
+                    return None;
+                }
+            };
+
+            // At this point src_branch and dst_branch match, so we only need to check one
+            let branch = match src_branch.match_indices('*').count() {
```

Comment:
nit: `.matches('*')` as we aren't interested in indices? or just convert `src_branch` to `StringPattern::glob()` unconditionally?

I'm not sure if unsupported glob chars are rejected by gix. If they are rejected by gix, we can just convert the `src_branch` pattern to `StringPattern::glob()`. If they aren't, we'll need to check `INVALID_REFSPEC_CHARS` here. Can you add test that includes `?` or `[foo]` in refspec?

Reply:
I confirmed that gix does reject all of these cases. I've added a snapshot test to validate we get a configuration error in that situation, and I've simplified the branch conversion to always use `StringPattern::glob`

Let me know if you have any further feedback on the extra test! Otherwise I'll put it on the merge queue tomorrow morning my time since you've already approved